
## This Notebook is a sort of tutorial for the beginners in Deep Learning and time-series data analysis. 

##  * The aim is just to show how to build the simplest Long Short-Term Memory (LSTM) recurrent neural network for the data.  
 
### * Jupyter notebook can be downloaded here: https://github.com/amirrezaeian/Individual-household-electric-power-consumption-Data-Set-

The description of data can be found here:
http://archive.ics.uci.edu/ml/datasets/Individual+household+electric+power+consumption

Attribute Information:
1.date: Date in format dd/mm/yyyy

2.time: time in format hh:mm:ss

3.global_active_power: household global minute-averaged active power (in kilowatt)

4.global_reactive_power: household global minute-averaged reactive power (in kilowatt)

5.voltage: minute-averaged voltage (in volt)

6.global_intensity: household global minute-averaged current intensity (in ampere)

7.sub_metering_1: energy sub-metering No. 1 (in watt-hour of active energy). It corresponds to the kitchen, containing mainly a dishwasher, an oven and a microwave (hot plates are not electric but gas powered).

8.sub_metering_2: energy sub-metering No. 2 (in watt-hour of active energy). It corresponds to the laundry room, containing a washing-machine, a tumble-drier, a refrigerator and a light.

9.sub_metering_3: energy sub-metering No. 3 (in watt-hour of active energy). It corresponds to an electric water-heater and an air-conditioner.

In [5]:
# Let`s import all packages that we may need:

import sys

import matplotlib.pyplot as plt  # this is used for the plot the graph
import numpy as np  # linear algebra
import pandas as pd  # data processing, CSV file I/O (e.g. pd.read_csv), data manipulation as in SQL
import seaborn as sns  # used for plot interactive graph.
from scipy.stats import randint
from sklearn import metrics  # for the check the error and accuracy of the model
from sklearn.feature_selection import SelectFromModel
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import KFold  # use for cross validation
from sklearn.model_selection import train_test_split  # to split the data into two parts
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import Pipeline  # pipeline making
from sklearn.preprocessing import StandardScaler  # for normalization
from sklearn.preprocessing import MinMaxScaler

""""
## for Deep-learing:
from keras.layers import Dense
from keras.models import Sequential
from keras.utils import to_categorical
from keras.optimizers import SGD 
from keras.callbacks import EarlyStopping
from keras.utils import np_utils
import itertools
from keras.layers import LSTM
from keras.layers.convolutional import Conv1D
from keras.layers.convolutional import MaxPooling1D
from keras.layers import Dropout
"""

'"\n## for Deep-learing:\nfrom keras.layers import Dense\nfrom keras.models import Sequential\nfrom keras.utils import to_categorical\nfrom keras.optimizers import SGD \nfrom keras.callbacks import EarlyStopping\nfrom keras.utils import np_utils\nimport itertools\nfrom keras.layers import LSTM\nfrom keras.layers.convolutional import Conv1D\nfrom keras.layers.convolutional import MaxPooling1D\nfrom keras.layers import Dropout\n'

In [7]:
## Data can be downloaded from: http://archive.ics.uci.edu/ml/machine-learning-databases/00235/
## Just open the zip file and grab the file 'household_power_consumption.txt' put it in the directory
## that you would like to run the code.


df = pd.read_csv(
    "household_power_consumption.txt",
    sep=";",
    parse_dates={"dt": ["Date", "Time"]},
    infer_datetime_format=True,
    low_memory=False,
    na_values=["nan", "?"],
    index_col="dt",
)

 ## 1) Note that data include 'nan' and '?' as a string. I converted both to numpy nan in importing stage (above) and treated both of them the same. 

## 2) I merged two columns 'Date' and 'Time' to 'dt'. 

## 3) I also converted in the above, the data to time-series type, by taking index to be the time. 

In [8]:
df.head()

Global_active_power  Global_reactive_power  Voltage  \
dt                                                                         
2006-12-16 17:24:00                4.216                  0.418   234.84   
2006-12-16 17:25:00                5.360                  0.436   233.63   
2006-12-16 17:26:00                5.374                  0.498   233.29   
2006-12-16 17:27:00                5.388                  0.502   233.74   
2006-12-16 17:28:00                3.666                  0.528   235.68   

                     Global_intensity  Sub_metering_1  Sub_metering_2  \
dt                                                                      
2006-12-16 17:24:00              18.4             0.0             1.0   
2006-12-16 17:25:00              23.0             0.0             1.0   
2006-12-16 17:26:00              23.0             0.0             2.0   
2006-12-16 17:27:00              23.0             0.0             1.0   
2006-12-16 17:28:00              15.8             0.0             1.0   

                     Sub_metering_3  
dt                                   
2006-12-16 17:24:00            17.0  
2006-12-16 17:25:00            16.0  
2006-12-16 17:26:00            17.0  
2006-12-16 17:27:00            17.0  
2006-12-16 17:28:00            17.0

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2075259 entries, 2006-12-16 17:24:00 to 2010-11-26 21:02:00
Data columns (total 7 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Global_active_power    float64
 1   Global_reactive_power  float64
 2   Voltage                float64
 3   Global_intensity       float64
 4   Sub_metering_1         float64
 5   Sub_metering_2         float64
 6   Sub_metering_3         float64
dtypes: float64(7)
memory usage: 126.7 MB


In [10]:
df.dtypes

Global_active_power      float64
Global_reactive_power    float64
Voltage                  float64
Global_intensity         float64
Sub_metering_1           float64
Sub_metering_2           float64
Sub_metering_3           float64
dtype: object

In [11]:
df.shape

(2075259, 7)

In [12]:
df.describe()

Global_active_power  Global_reactive_power       Voltage  \
count         2.049280e+06           2.049280e+06  2.049280e+06   
mean          1.091615e+00           1.237145e-01  2.408399e+02   
std           1.057294e+00           1.127220e-01  3.239987e+00   
min           7.600000e-02           0.000000e+00  2.232000e+02   
25%           3.080000e-01           4.800000e-02  2.389900e+02   
50%           6.020000e-01           1.000000e-01  2.410100e+02   
75%           1.528000e+00           1.940000e-01  2.428900e+02   
max           1.112200e+01           1.390000e+00  2.541500e+02   

       Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
count      2.049280e+06    2.049280e+06    2.049280e+06    2.049280e+06  
mean       4.627759e+00    1.121923e+00    1.298520e+00    6.458447e+00  
std        4.444396e+00    6.153031e+00    5.822026e+00    8.437154e+00  
min        2.000000e-01    0.000000e+00    0.000000e+00    0.000000e+00  
25%        1.400000e+00    0.000000e+00    0.000000e+00    0.000000e+00  
50%        2.600000e+00    0.000000e+00    0.000000e+00    1.000000e+00  
75%        6.400000e+00    0.000000e+00    1.000000e+00    1.700000e+01  
max        4.840000e+01    8.800000e+01    8.000000e+01    3.100000e+01

In [13]:
df.columns

Index(['Global_active_power', 'Global_reactive_power', 'Voltage',
       'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3'],
      dtype='object')

## Dealing with missing values  'nan' with a test statistic

In [14]:
## finding all columns that have nan:

droping_list_all = []
for j in range(0, 7):
    if not df.iloc[:, j].notnull().all():
        droping_list_all.append(j)
        # print(df.iloc[:,j].unique())
droping_list_all

[0, 1, 2, 3, 4, 5, 6]

In [15]:
# filling nan with mean in any columns

for j in range(0, 7):
    df.iloc[:, j] = df.iloc[:, j].fillna(df.iloc[:, j].mean())

In [16]:
# another sanity check to make sure that there are not more any nan
df.isnull().sum()

Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
dtype: int64

# Data visualization

## * Below I resample over day, and show the sum and mean of Global_active_power. It is seen that mean and sum of resampled data set, have similar structure.

In [ ]:
df.Global_active_power.resample("D").sum().plot(
    title="Global_active_power resampled over day for sum"
)
# df.Global_active_power.resample('D').mean().plot(title='Global_active_power resampled over day', color='red')
plt.tight_layout()
plt.show()

df.Global_active_power.resample("D").mean().plot(
    title="Global_active_power resampled over day for mean", color="red"
)
plt.tight_layout()
plt.show()

In [ ]:
### Below I show mean and std of 'Global_intensity' resampled over day
r = df.Global_intensity.resample("D").agg(["mean", "std"])
r.plot(subplots=True, title="Global_intensity resampled over day")
plt.show()

In [ ]:
### Below I show mean and std of 'Global_reactive_power' resampled over day
r2 = df.Global_reactive_power.resample("D").agg(["mean", "std"])
r2.plot(subplots=True, title="Global_reactive_power resampled over day", color="red")
plt.show()

In [ ]:
### Sum of 'Global_active_power' resampled over month
# Sum of 'Global_active_power' resampled over month
df["Global_active_power"].resample("M").mean().plot(kind="bar")
plt.xticks(rotation=60)
plt.ylabel("Global_active_power")
plt.title("Global_active_power per month (averaged over month)")
plt.show()

In [ ]:
## Mean of 'Global_active_power' resampled over quarter
df["Global_active_power"].resample("Q").mean().plot(kind="bar")
plt.xticks(rotation=60)
plt.ylabel("Global_active_power")
plt.title("Global_active_power per quarter (averaged over quarter)")
plt.show()

### It is very important to note from above two plots that resampling over larger time inteval, will diminish the periodicity of system as we expect. This is important for machine learning feature engineering. 

In [ ]:
## mean of 'Voltage' resampled over month
df["Voltage"].resample("M").mean().plot(kind="bar", color="red")
plt.xticks(rotation=60)
plt.ylabel("Voltage")
plt.title("Voltage per quarter (summed over quarter)")
plt.show()

In [ ]:
df["Sub_metering_1"].resample("M").mean().plot(kind="bar", color="brown")
plt.xticks(rotation=60)
plt.ylabel("Sub_metering_1")
plt.title("Sub_metering_1 per quarter (summed over quarter)")
plt.show()

 ## * It is seen from the above plots that the mean of 'Volage' over month is pretty much constant compared to other features. This is important again in feature selection.

In [ ]:
# Below I compare the mean of different featuresresampled over day.
# specify columns to plot
cols = [0, 1, 2, 3, 5, 6]
i = 1
groups = cols
values = df.resample("D").mean().values
# plot each column
plt.figure(figsize=(15, 10))
for group in groups:
    plt.subplot(len(cols), 1, i)
    plt.plot(values[:, group])
    plt.title(df.columns[group], y=0.75, loc="right")
    i += 1
plt.show()

In [ ]:
## resampling over week and computing mean
df.Global_reactive_power.resample("W").mean().plot(color="y", legend=True)
df.Global_active_power.resample("W").mean().plot(color="r", legend=True)
df.Sub_metering_1.resample("W").mean().plot(color="b", legend=True)
df.Global_intensity.resample("W").mean().plot(color="g", legend=True)
plt.show()

In [ ]:
# Below I show hist plot of the mean of different feature resampled over month
df.Global_active_power.resample("M").mean().plot(kind="hist", color="r", legend=True)
df.Global_reactive_power.resample("M").mean().plot(kind="hist", color="b", legend=True)
# df.Voltage.resample('M').sum().plot(kind='hist',color='g', legend=True)
df.Global_intensity.resample("M").mean().plot(kind="hist", color="g", legend=True)
df.Sub_metering_1.resample("M").mean().plot(kind="hist", color="y", legend=True)
plt.show()

In [ ]:
## The correlations between 'Global_intensity', 'Global_active_power'
data_returns = df.pct_change()
sns.jointplot(x="Global_intensity", y="Global_active_power", data=data_returns)

plt.show()

### * From above two plots it is seen that 'Global_intensity' and 'Global_active_power' correlated. But 'Voltage', 'Global_active_power' are less correlated. This is important observation for machine learning purpose. 

In [ ]:
## The correlations between 'Voltage' and  'Global_active_power'
sns.jointplot(x="Voltage", y="Global_active_power", data=data_returns)
plt.show()

# Correlations among features

In [ ]:
# Correlations among columns
plt.matshow(df.corr(method="spearman"), vmax=1, vmin=-1, cmap="PRGn")
plt.title("without resampling", size=15)
plt.colorbar()
plt.show()

In [ ]:
# Correlations of mean of features resampled over months


plt.matshow(
    df.resample("M").mean().corr(method="spearman"), vmax=1, vmin=-1, cmap="PRGn"
)
plt.title("resampled over month", size=15)
plt.colorbar()
plt.margins(0.02)
plt.matshow(
    df.resample("A").mean().corr(method="spearman"), vmax=1, vmin=-1, cmap="PRGn"
)
plt.title("resampled over year", size=15)
plt.colorbar()
plt.show()

## * It is seen from above that with resampling techniques one can change the correlations among features. This is important for feature  engineering.

# Machine-Leaning: LSTM Data Preparation and feature engineering

### * I will apply recurrent nueral network (LSTM) which is best suited for time-seriers and sequential problem. This approach is the best if we have large data.  

### * I will frame the supervised learning problem as predicting the Global_active_power at the current time (t) given the Global_active_power measurement and other features at the prior time step.

In [17]:
def series_to_supervised(data, n_in=1, n_out=1, dropnan=True):
    n_vars = 1 if type(data) is list else data.shape[1]
    dff = pd.DataFrame(data)
    cols, names = list(), list()
    # input sequence (t-n, ... t-1)
    for i in range(n_in, 0, -1):
        cols.append(dff.shift(i))
        names += [("var%d(t-%d)" % (j + 1, i)) for j in range(n_vars)]
    # forecast sequence (t, t+1, ... t+n)
    for i in range(0, n_out):
        cols.append(dff.shift(-i))
        if i == 0:
            names += [("var%d(t)" % (j + 1)) for j in range(n_vars)]
        else:
            names += [("var%d(t+%d)" % (j + 1, i)) for j in range(n_vars)]
    # put it all together
    agg = pd.concat(cols, axis=1)
    agg.columns = names
    # drop rows with NaN values
    if dropnan:
        agg.dropna(inplace=True)
    return agg

### * In order to reduce the computation time, and also get a quick result to test the model.  One can resmaple the data over hour (the original data are given in minutes). This will reduce the size of data from 2075259 to 34589 but keep the overall strucure of data as shown in the above.   

In [18]:
## resampling of data over hour
df_resample = df.resample("h").mean()
df_resample.shape

(34589, 7)

In [28]:
## * Note: I scale all features in range of [0,1].

## If you would like to train based on the resampled data (over hour), then used below
values = df_resample.values


## full data without resampling
# values = df.values

# integer encode direction
# ensure all data is float
# values = values.astype('float32')
# normalize features
scaler = MinMaxScaler(feature_range=(0, 1))
scaled = scaler.fit_transform(values)
# frame as supervised learning
reframed = series_to_supervised(scaled, 1, 1)
print(reframed.head())
# drop columns we don't want to predict
reframed.drop(reframed.columns[[8, 9, 10, 11, 12, 13]], axis=1, inplace=True)
print(reframed.head())
# print(scaled)
# print(df_resample.head())

   var1(t-1)  var2(t-1)  var3(t-1)  var4(t-1)  var5(t-1)  var6(t-1)  \
1   0.636816   0.295738   0.337945   0.631157        0.0   0.011366   
2   0.545045   0.103358   0.335501   0.541487        0.0   0.144652   
3   0.509006   0.110073   0.283802   0.502152        0.0   0.030869   
4   0.488550   0.096987   0.315987   0.481110        0.0   0.000000   
5   0.455597   0.099010   0.434417   0.449904        0.0   0.008973   

   var7(t-1)   var1(t)   var2(t)   var3(t)   var4(t)  var5(t)   var6(t)  \
1   0.782418  0.545045  0.103358  0.335501  0.541487      0.0  0.144652   
2   0.782676  0.509006  0.110073  0.283802  0.502152      0.0  0.030869   
3   0.774169  0.488550  0.096987  0.315987  0.481110      0.0  0.000000   
4   0.778809  0.455597  0.099010  0.434417  0.449904      0.0  0.008973   
5   0.798917  0.322555  0.072536  0.495847  0.323529      0.0  0.002872   

    var7(t)  
1  0.782676  
2  0.774169  
3  0.778809  
4  0.798917  
5  0.205723  
   var1(t-1)  var2(t-1)  var3(t-1)  va

In [20]:
reframed.head()

var1(t-1)  var2(t-1)  var3(t-1)  var4(t-1)  var5(t-1)  var6(t-1)  \
1   0.636816   0.295738   0.337945   0.631157        0.0   0.011366   
2   0.545045   0.103358   0.335501   0.541487        0.0   0.144652   
3   0.509006   0.110073   0.283802   0.502152        0.0   0.030869   
4   0.488550   0.096987   0.315987   0.481110        0.0   0.000000   
5   0.455597   0.099010   0.434417   0.449904        0.0   0.008973   

   var7(t-1)   var1(t)  
1   0.782418  0.545045  
2   0.782676  0.509006  
3   0.774169  0.488550  
4   0.778809  0.455597  
5   0.798917  0.322555

### * Above I showed 7 input variables (input series) and the 1 output variable for 'Global_active_power' at the current  time in hour (depending on resampling). 

# Splitting the rest of data to train and validation sets

### * First, I split the prepared dataset into train and test sets. To speed up the training of the model (for the sake of the demonstration), we will only train the model on the first year of data, then evaluate it on the next 3 years of data.

In [29]:
# split into train and test sets
values = reframed.values

n_train_time = 365 * 24
train = values[:n_train_time, :]
test = values[n_train_time:, :]
##test = values[n_train_time:n_test_time, :]
# split into input and outputs
train_X, train_y = train[:, :-1], train[:, -1]
test_X, test_y = test[:, :-1], test[:, -1]
# reshape input to be 3D [samples, timesteps, features]
train_X = train_X.reshape((train_X.shape[0], 1, train_X.shape[1]))
test_X = test_X.reshape((test_X.shape[0], 1, test_X.shape[1]))
print(train_X.shape, train_y.shape, test_X.shape, test_y.shape)
# We reshaped the input into the 3D format as expected by LSTMs, namely [samples, timesteps, features].

(8760, 1, 7) (8760,) (25828, 1, 7) (25828,)


# Model architecture

### 1)  LSTM with 100 neurons in the first visible layer 
### 3) dropout 20%
### 4) 1 neuron in the output layer for predicting Global_active_power. 
### 5) The input shape will be 1 time step with 7 features.

### 6) I use the Mean Absolute Error (MAE) loss function and the efficient Adam version of stochastic gradient descent.
### 7) The model will be fit for 20 training epochs with a batch size of 70.


In [ ]:
model = Sequential()
model.add(LSTM(100, input_shape=(train_X.shape[1], train_X.shape[2])))
model.add(Dropout(0.2))
#    model.add(LSTM(70))
#    model.add(Dropout(0.3))
model.add(Dense(1))
model.compile(loss="mean_squared_error", optimizer="adam")


# fit network
history = model.fit(
    train_X,
    train_y,
    epochs=20,
    batch_size=70,
    validation_data=(test_X, test_y),
    verbose=2,
    shuffle=False,
)

# summarize history for loss
plt.plot(history.history["loss"])
plt.plot(history.history["val_loss"])
plt.title("model loss")
plt.ylabel("loss")
plt.xlabel("epoch")
plt.legend(["train", "test"], loc="upper right")
plt.show()

# make a prediction
yhat = model.predict(test_X)
test_X = test_X.reshape((test_X.shape[0], 7))
# invert scaling for forecast
inv_yhat = np.concatenate((yhat, test_X[:, -6:]), axis=1)
inv_yhat = scaler.inverse_transform(inv_yhat)
inv_yhat = inv_yhat[:, 0]
# invert scaling for actual
test_y = test_y.reshape((len(test_y), 1))
inv_y = np.concatenate((test_y, test_X[:, -6:]), axis=1)
inv_y = scaler.inverse_transform(inv_y)
inv_y = inv_y[:, 0]
# calculate RMSE
rmse = np.sqrt(mean_squared_error(inv_y, inv_yhat))
print("Test RMSE: %.3f" % rmse)

### Note that in order to improve the model, one has to adjust epochs and batch_size.

In [ ]:
## time steps, every step is one hour (you can easily convert the time step to the actual time index)
## for a demonstration purpose, I only compare the predictions in 200 hours.

aa = [x for x in range(200)]
plt.plot(aa, inv_y[:200], marker=".", label="actual")
plt.plot(aa, inv_yhat[:200], "r", label="prediction")
plt.ylabel("Global_active_power", size=15)
plt.xlabel("Time step", size=15)
plt.legend(fontsize=15)
plt.show()

# Final remarks

### * Here I have used the LSTM neural network which is now the state-of-the-art for sequencial problems. 

### * In order to reduce the computation time, and get some results quickly, I took the first year of data (resampled over hour) to train the model and the rest of data to test the model.  

### * I put together a very simple LSTM neural-network to show that one can obtain reasonable predictions. However numbers of rows is too high and as a result the computation is very time-consuming (even for the simple model in the above it took few mins to be run on  2.8 GHz Intel Core i7).  The Best is to write the last part of code using Spark (MLlib) running on GPU.  

### * Moreover, the neural-network architecture that I have designed is a toy model. It can be easily improved by adding CNN  and dropout layers.  The CNN is useful here since there are correlations in data (CNN layer is a good way to probe the local structure of data).   

In [ ]:
class LSTM:
    def __init__(self, data):
        self.data = np.reshape(data, (-1, 1))

    def scale_data(self, scaler=MinMaxScaler(feature_range=(0, 1))):
        self.data = scaler.fit_transform(self.data)

    def split_data(self, split=0.75):
        train_len = int(len(self.data) * split)
        test_len = len(self.data) - train_len
        self.train, self.test = (
            self.data[0:train_len, :],
            self.data[train_len : len(self.data), :],
        )
        return self.train, self.test

    def make_lags(self, list_of_data, lags, lead_time=1):
        return pd.concat(
            {f"y_step_{i + 1}": ts.shift(-i) for i in range(steps)}, axis=1
        )

    def xy_split(self, data_to_split):
        return np.array(data_to_split.iloc[:, :-1]), np.array(data_to_split.iloc[:, -1])

    def reshape_to_3(self, data_to_reshape):
        return np.reshape(
            data_to_reshape, (data_to_reshape.shape[0], 1, data_to_reshape.shape[1])
        )

    def store_values(self, x_train, x_test, y_train, y_test):
        self.x_train = x_train
        self.x_test = x_test
        self.y_train = y_train
        self.y_test = y_test

    def create_model(self, LSTM_neurons=100, dropout_percent=0.2):
        self.model = keras.Sequential()
        self.model.add(
            layers.LSTM(
                LSTM_neurons, input_shape=(self.x_train.shape[1], self.x_train.shape[2])
            )
        )
        self.model.add(layers.Dropout(dropout_percent))
        self.model.add(layers.Dense(1))
        self.model.compile(loss="mean_squared_error", optimizer="adam")

    def fit_model(
        epochs,
        batch_size,
    ):
        self.history = self.model.fit(
            self.x_train,
            self.y_train,
            epochs=epochs,
            batch_size=batch_size,
            validation_data=(self.x_test, self.y_test),
            callbacks=[keras.callbacks.EarlyStopping(monitor="val_loss", patience=10)],
            verbose=1,
            shuffle=False,
        )

        print(self.model.summary())

    def predict_evaluate(self):
        self.train_predict = self.model.predict(self.x_train)
        self.test_predict = self.model.predict(self.x_test)
        # invert predictions back to previous values  
        self.train_predict = scaler.inverse_transform(self.train_predict)
        self.y_train = scaler.inverse_transform([self.y_train])
        self.test_predict = scaler.inverse_transform(self.test_predict)
        self.y_test = scaler.inverse_transform([self.y_test])
        
        mae = mean_absolute_error(self.y_train[0], self.train_predict[:, 0])
        rmse = np.sqrt(mean_squared_error(self.Y_train[0],self. train_predict[:, 0]))
        tmse = mean_absolute_error(self.y_test[0], self.test_predict[:, 0])
        trmse = np.sqrt(mean_squared_error(self.Y_test[0], self.test_predict[:, 0]))

        print("Train Mean Absolute Error: %.3f" % mae)
        print("Train Root Mean Squared Error: %.3f" % rmse)
        print("Test Mean Absolute Error: %.3f" % tmse)
        print("Test Root Mean Squared Error: %.3f" % trmse)
    def plot_loss(self):
        plt.figure(figsize=(10, 5))
        plt.plot(history.history["loss"], label="Train Loss")
        plt.plot(history.history["val_loss"], label="Test Loss")
        plt.title("model loss")
        plt.ylabel("loss")
        plt.xlabel("epochs")
        plt.legend(loc="upper right")
        plt.show();
    def plot_graph(self, timestep, window_start,window_end ,data_col_name):
        
        time_range  = [i for i in range(timestep)]
        plt.figure(figsize=(8, 4))
        plt.plot(time_range , self.y_test[0][window_start:window_end], color='b',marker=".", label="Reality")
        plt.plot(time_range , self.test_predict[:, 0][window_start:windo], "r", label="Prediction")
        # plt.tick_params(left=False, labelleft=True) #remove ticks
        plt.tight_layout()
        sns.despine(top=True)
        plt.subplots_adjust(left=0.07)
        plt.ylabel(data_col_name, size=15)
        plt.xlabel("Time step", size=15)
        plt.legend(fontsize=15)
        plt.show();
              